open-policy-agent/service-account-mgmt/unique_serviceaccount_per_pod/template.yaml

```diff
@@ -0,0 +1,36 @@
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: k8spoduniqueserviceaccountvalidation
+  annotations:
+    description: Checks wheter the pod has a unique serviceAccount, correlated with the pod's name
+spec:
+  crd:
+    spec:
+      names:
+        kind: K8sPodUniqueServiceAccountValidation
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package K8sPodUniqueServiceAccountValidation
+
+        missing(obj, field) = true {
+          not obj[field]
+        }
+        
+        missing(obj, field) = true {
+          obj[field] == ""
+        }
+
+        violation[{"msg": msg}] {
+          pod := input.review.object
+
+          missing(pod.spec, "serviceAccountName")
+
+          msg := "serviceAccountName is required in the pod spec"
+        }
+
+        violation[{"msg": msg}] {
```

Comment:
**DISCUSSION** @tommeramber @igorkchyts 

Is it practical in your opinion to validate the `serviceaccount` name against the `pod` name?
A pod is usually generated with a random `id` -> `<podname>-<replicaset id>-<pod id>`. The id will change with deployment rollouts.

When trying to use the policy with a `deployment` object -

- Tried to deploy the next object.
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-openshift
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello-openshift
  template:
    metadata:
      labels:
        app: hello-openshift
    spec:
      serviceAccountName: hello-openshift
      containers:
      - name: hello-openshift
        image: openshift/hello-openshift:latest
        ports:
        - containerPort: 80
```
And received the next error -
"
replicaset-controller  Error creating: admission webhook "validation.gatekeeper.sh" denied the request: [denied by pod-unique-serviceaccount-validation] serviceAccountName should be the same as the name of the pod: hello-openshift-5f69b7bc46-9t2hc. In case you do not have such an serviceAccount already, you are required to generate one.
"

I'm receiving the error because the service account name:  `hello-openshift` is not equal to the pod: `hello-openshift-5f69b7bc46-h8sfs`

In my opinion, It would be better to evaluate the `serviceaccount` name against a more high level object, with a persistent name like `deployment`. What do you think?

Reply:
@michaelkotelnikov @tommeramber 
Indeed we can't validate service account against pod name. But, we need to validate each pod serviceaccount against deployment, statefulset, daemonset, job, deploymentconfig. 

But, in this case pod object can be created manually by the user and any serviceAccount can be assigned to this pod anyway. 

Another idea is to validate against labels, but then we enforce cluster users to attach specific label to each pod or deployment object that contains service account name.

In my opinion it is also reasonable that the policy will not allow pods that use default service accounts in the namespace, this means that the user must create service account for its services and the service account is governed by the scc that are applied.